### DISCUSSION TITLE
Music Sheet for MIDI

### DISCUSSION THREAD

**Discussion by Treata11** [CoderLine/alphaTab#1315] [Q&A] [answered]
Is it possible to load and display music sheets from standard Midi files, using alphaTab?

Is there possibly a framework in C family languages, capable of converting MIDIs to data sources like Guitar Pro or the built in markup language (alphaTex)?

**Comment by Danielku15** [comment, marked as answer]
> Is it possible to load and display music sheets from standard Midi files, using alphaTab?

Technically is is possible to add support for midi and we have it on our roadmap. As of today we have no built-in support. https://github.com/orgs/CoderLine/projects/12?pane=issue&itemId=21406500

There are some specifics to importing midi which make it maybe a bit more complicated: 

* There are 3 types of midi files specified in the SMF 1.0 standard describing how the individual tracks are organized. 
  * Type 0 (multi-channel track): We have to make assumptions llike that every channel becomes 1 track. 
  * Type 1 (multi track - simultaneous)
  * Type 2 (multi track - sequential)
* There might be some metadata like the track name which we should try to decode. 
* We have to translate the raw midi note durations between note-on and note-off into proper beat durations (full, half, quarter,..) and might need to connect multiple beats . 


> Is there possibly a framework in C family languages, capable of converting MIDIs to data sources like Guitar Pro or the built in markup language (alphaTex)?

I am not aware of any framework or library offering this but also didn't do much research on the topic. I know Guitar Pro can import midi itself but its not scriptable or usable as a library. 

While we currently only have some basic support for MusicXML you could try to go a two step approach: 
* Convert MIDI to MusicXML (there are quite some tools and libs around for that)
* Let alphaTab import MusicXML (we might be able to do some bugfixing if something doesn't work nicely)
